sklearnex/covariance/incremental_covariance.py

```diff
@@ -0,0 +1,131 @@
+
+from daal4py.sklearn._device_offload import support_usm_ndarray
+from daal4py.sklearn._utils import control_n_jobs, run_with_n_jobs
+from onedal.covariance import (
+    IncrementalEmpiricalCovariance as onedal_IncrementalEmpiricalCovariance,
+)
+
+
+@control_n_jobs
+class IncrementalEmpiricalCovariance:
+    """Incremental estimator for covariance.
+    Allows to compute empirical covariance estimated by maximum
+    likelihood method if data are splitted into batches.
+
+    Parameters
+    ----------
+    batch_size : int, default=None
+        The number of samples to use for each batch. Only used when calling
+        ``fit``. If ``batch_size`` is ``None``, then ``batch_size``
+        is inferred from the data and set to ``5 * n_features``, to provide a
+        balance between approximation accuracy and memory consumption.
+
+    Attributes
+    ----------
+    location_ : ndarray of shape (n_features,)
+        Estimated location, i.e. the estimated mean.
+
+    covariance_ : ndarray of shape (n_features, n_features)
+        Estimated covariance matrix
+
+    """
+
+    _onedal_incremental_covariance = staticmethod(onedal_IncrementalEmpiricalCovariance)
+
+    def __init__(self, batch_size=None):
+        self._need_to_finalize = False  # If True then finalize compute should
+        #      be called to obtain covariance_ or location_ from partial compute data
+        self.batch_size = batch_size
+
+    def _onedal_finalize_compute(self):
+        assert hasattr(self, "_onedal_estimator")
+        self._onedal_estimator.finalize_compute()
+        self._need_to_finalize = False
+
+    def _onedal_partial_compute(self, X):
+        onedal_params = {
+            "method": "dense",
+            "bias": True,
+        }
+        if not hasattr(self, "_onedal_estimator"):
+            self._onedal_estimator = self._onedal_incremental_covariance(**onedal_params)
+        self._onedal_estimator.partial_compute(X)
+        self._need_to_finalize = True
+
+    @property
+    def covariance_(self):
+        if self._need_to_finalize:
+            self._onedal_finalize_compute()
+        return self._onedal_estimator.covariance_
+
+    @property
+    def location_(self):
+        if self._need_to_finalize:
+            self._onedal_finalize_compute()
+        return self._onedal_estimator.location_
+
+    @support_usm_ndarray()
+    @run_with_n_jobs
+    def partial_fit(self, X):
+        """Incremental fit with X. All of X is processed as a single batch.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            Training data, where `n_samples` is the number of samples and
+            `n_features` is the number of features.
+
+        Returns
+        -------
+        self : object
+            Returns the instance itself.
+        """
+        X = check_array(X, dtype=[np.float64, np.float32])
+        self._onedal_partial_compute(X)
+        return self
+
+    @support_usm_ndarray()
```

Comment:
what does support_usm_ndarray do to these functions?

Reply:
it takes queue from data dpctl/dpnp container and put it as queue to the method args